core/genesis/file.go

```diff
@@ -0,0 +1,261 @@
+package genesis
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+type FileGenesisSource struct {
+	sourceDir  string
+	moduleName string
+
+	// the RawMessage from the genesis.json app_state.<module> that got passed into InitChain
+	moduleRootJson json.RawMessage
+}
+
+const (
+	fileOpenflag  = os.O_CREATE | os.O_WRONLY
+	flieOpenMode  = fs.FileMode(0o600)
+	dirCreateMode = fs.FileMode(0o700)
+)
+
+// NewFileGenesisSource returns a new GenesisSource for the provided
+// source directory and the provided module name where it is assumed
+// that it contains encoded JSON data in the file or in the moduleState
+// of the appState be passed from RequestInitChain.
+func NewFileGenesisSource(sourceDir, moduleName string, rawModuleState json.RawMessage) GenesisSource {
+	return &FileGenesisSource{
+		sourceDir:      filepath.Clean(sourceDir),
+		moduleName:     moduleName,
+		moduleRootJson: rawModuleState,
+	}
+}
+
+// OpenReader opens the source field reading from the given parameters,
+// and returns a ReadCloser.
+// It will try to open the field in order following by:
+// <sourceDir>/<module>/<field>.json
+// <field> key inside <sourceDir>/<module>.json
+// app_state.<module>.<field> key from moduleRootJson
+func (f *FileGenesisSource) OpenReader(field string) (io.ReadCloser, error) {
+	// try reading genesis data from <sourceDir>/<module>/<field>.json
+	fName := fmt.Sprintf("%s.json", field)
+	fPath := filepath.Join(f.sourceDir, f.moduleName)
+
+	fp, err := os.Open(filepath.Clean(filepath.Join(fPath, fName)))
+	if err == nil {
+		return fp, nil
+	}
+
+	if !os.IsNotExist(err) {
+		return nil, fmt.Errorf("unexpected error: %w", err)
+	}
+
+	// if cannot find it, try reading from <sourceDir>/<module>.json
+	rawBz, err := f.ReadRawJSON()
+	if err == nil {
+		f.moduleRootJson = rawBz
```

Comment:
I'm not sure it's best to overwrite the genesis that came from initchain here. Maybe they should actually be treated with merge semantics if both exist.

Reply:
Also, can we do it easily without `unmarshal` and `marshal` data? it could be time-consuming.

Reply:
Not sure about the use case of putting the module parameters into the `app_state` of genesis.json and putting the module state to `module.json`. Is there any usage the SDK doesn't want(or need) to load the module state and execute the app?

Reply:
added the module state merge in `ReadRawJSON`, and also a new flag for avoiding duplicated read if the dev tries calling a different module field.